charmhelpers/fetch/ubuntu.py

```diff
@@ -189,17 +190,185 @@
+        previous = None
+        pkg = {}
+        for line in output.splitlines():
+            if not line:
+                if 'package' in pkg:
+                    pkgs.update({pkg['package']: pkg})
+                    pkg = {}
+                continue
+            if line.startswith(' '):
+                if previous and previous in pkg:
+                    pkg[previous] += os.linesep + line.lstrip()
+                continue
+            if ':' in line:
+                kv = line.split(':', 1)
+                key = kv[0].lower()
+                if key == 'n':
+                    continue
+                previous = key
+                pkg.update({key: kv[1].lstrip()})
+    except subprocess.CalledProcessError as cp:
+        # ``apt-cache`` returns 100 if none of the packages asked for exist in
+        # the apt cache.
+        if cp.returncode != 100:
+            raise
+    return pkgs
+
+
+def upstream_version(version):
+    """Extracts upstream version from a version string.
+
+    Note that one of the main purposes for this function is to avoid dependency
+    on the ``python-apt`` package.
+
+    The indicated package is a wrapper around the ``apt`` C++ library which is
+    tightly connected to the version of the distribution it is shipped on.
+    This in turn makes it incredibly hard to distribute as a wheel for a charm
+    that supports a large span of distro versions.
+
+    :param version: Version string
+    :type version: str
+    :returns: Upstream version
+    :rtype: str
+
+    Upstream reference: https://salsa.debian.org/apt-team/apt/blob/master/
+                                apt-pkg/deb/debversion.cc#L259
+    """
+    if version:
+        version = version[version.find(':') + 1:]
+        if '-' in version:
+            version = version[:version.find('-')]
+    return version
+
+
+def version_compare(candidate, current):
+    """Compare the given versions.
+
+    Note that one of the main purposes for this function is to avoid dependency
+    on the ``python-apt`` package.
+
+    The indicated package is a wrapper around the ``apt`` C++ library which is
+    tightly connected to the version of the distribution it is shipped on.
+    This in turn makes it incredibly hard to distribute as a wheel for a charm
+    that supports a large span of distro versions.
+
+    :param candidate: version string
+    :type candidate: str
+    :param current: version string
+    :type current: str
+    :returns: False if candidate <= current, True if candidate > current
+    :raises: subprocess.CalledProcessError
+    """
+    try:
+        subprocess.check_call(['dpkg', '--compare-versions', candidate, 'gt',
+                               current],
+                              stderr=subprocess.STDOUT,
+                              universal_newlines=True)
+        return True
+    except subprocess.CalledProcessError as cp:
+        if cp.returncode == 1:
+            return False
+        raise
```

Comment:
Do we really want to `raise` here?  i.e. wouldn't an error be `False` anyway?

Reply:
Yes, if the system or the ``dpkg`` binary causes a unexpected error we would want to know about it. True/False is only valid when we can determine the relationship between the versions.  A updated version is on the way where this function restores it's ternary result (for >, = or <)